api-admin/src/main/java/bio/terra/pearl/api/admin/service/notifications/NotificationConfigExtService.java

```diff
@@ -70,19 +98,60 @@ public NotificationConfig replace(
-    if (update.getEmailTemplate() != null && update.getEmailTemplateId() == null) {
-      // this is a new email template, set the portal appropriately
-      update.getEmailTemplate().setPortalId(portalEnvironment.getPortalId());
-    }
-    NotificationConfig newConfig = notificationConfigService.create(update);
-
-    // once the new creation succeeds, deactivate the prior config.
+    authNotificationConfig(existing, portalEnvironment, studyEnvironment);
+    NotificationConfig newConfig = create(update, studyEnvironment, portalEnvironment);
+    // after creating the new config, deactivate the old config
     existing.setActive(false);
     notificationConfigService.update(existing);
     return newConfig;
   }
+
+  /**
+   * deactivates the notification config with configId, and adds a new config as specified in the
+   * update object. Note though that the portalEnvironmentId and studyEnvironmentId will be set from
+   * the portalShortcode and studyShortcode params. If the update contains a new email template,
+   * that template will be created as well.
+   */
+  @Transactional
+  public NotificationConfig create(
+      String portalShortcode,
+      String studyShortcode,
+      EnvironmentName environmentName,
+      NotificationConfig newConfig,
+      AdminUser user) {
+    authUtilService.authUserToPortal(user, portalShortcode);
+    PortalEnvironment portalEnvironment =
+        portalEnvironmentService.findOne(portalShortcode, environmentName).get();
+    authUtilService.authUserToStudy(user, portalShortcode, studyShortcode);
+    StudyEnvironment studyEnvironment =
+        studyEnvironmentService.findByStudy(studyShortcode, environmentName).get();
+
+    return create(newConfig, studyEnvironment, portalEnvironment);
+  }
+
+  /** confirms the given config is associated with the given study and portal environments */
+  private void authNotificationConfig(
+      NotificationConfig config,
+      PortalEnvironment portalEnvironment,
+      StudyEnvironment studyEnvironment) {
+    if (!studyEnvironment.getId().equals(config.getStudyEnvironmentId())
+        || !portalEnvironment.getId().equals(config.getPortalEnvironmentId())) {
+      throw new IllegalArgumentException("config does not match the study and portal environment");
+    }
+  }
+
+  private NotificationConfig create(
+      NotificationConfig newConfig,
+      StudyEnvironment studyEnvironment,
+      PortalEnvironment portalEnvironment) {
+    newConfig.cleanForCopying();
+    newConfig.setStudyEnvironmentId(studyEnvironment.getId());
+    newConfig.setPortalEnvironmentId(portalEnvironment.getId());
+    if (newConfig.getEmailTemplate() != null && newConfig.getEmailTemplateId() == null) {
+      // this is a new email template, set the portal appropriately
+      newConfig.getEmailTemplate().setPortalId(portalEnvironment.getPortalId());
+    }
+    NotificationConfig savedConfig = notificationConfigService.create(newConfig);
```

Comment:
I know some style guides frown on this, for me these last-line assignments are an extremely useful/convenient place to put a debug breakpoint -- that way I can quickly inspect what was already returned.  So that's what you're seeing here.

Reply:
You are correct.  Sonar, among others, flag this pattern, so it stands out.
But your reason for use makes sense to me.